ref(userInfo.email): functionality.

hristoterezov opened the pull request.

virtuacoplenny commented:
What was the intend of this refactor? The commit message/description could be more descriptive. I see that you moved the setting of user.email, but why did you move it? It made more sense to be on the user or did it fix some bug? Is the circular dependency fix related to this change?

hristoterezov commented:
@virtuacoplenny Note that there aren't any functional changes in the PR. The userInfo.email is still set to the local participant as it was before. 

What is changed:
1) The imports are fixed. There was a problem with circular dependancies before and that's why the imports where using not the feature directory but directly the files. The circular dependancy  still exists but it is not a problem anymore.
2) Setting the email on the local participant from url was moved to the middleware of base/participants from base/settings.

I'm sorry the commit message is not clear!

Here's the whole story:
I'm working on using another prop from the userInfo object. My first implementation was to use the same approach as the email so I just added the prop into `_updateLocalParticipantFromUrl`. Then I decided I will fix the problem with the circular dependancies. And then I decided that since this function has nothing to do with settings it will be better to move it to base/participants. After that I decided that the new functionality that I implement doesn't fit very well in the local participant and I moved it elsewhere. But I figured it will be still good to commit the refactoring I've already done for the userInfo.email.

How do you want me to change the commit message / description? Do you want me to change anything in the implementation?

hristoterezov commented:
> Yes, before merging/rebasing you should update the commit message a bit to express what the change is. If you feel necessary, the description should also include intent. I don't see a need for logic changes.

How about "ref(userInfo.email): move to base/participants"?

virtuacoplenny commented:
> > Yes, before merging/rebasing you should update the commit message a bit to express what the change is. If you feel necessary, the description should also include intent. I don't see a need for logic changes.
> 
> How about "ref(userInfo.email): move to base/participants"?

Okay. That's better to me. A short sentence about why you moved it might be useful too, but I leave it to your discretion.

saghul commented on react/features/base/devices/middleware.js:
```diff
@@ -47,6 +31,25 @@ const JITSI_TRACK_ERROR_TO_MESSAGE_KEY_MAP = {
  */
 // eslint-disable-next-line no-unused-vars
 MiddlewareRegistry.register(store => next => action => {
+    // If defined in the global scope, it causes ReferenceError for JitsiTrackErrors. It seems that circular
```
Please don't do this. Fix the circular dependency error instead of introduccing a hack.

hristoterezov replied:
What do you mean "Fix the circular dependency error"? There's no "circular dependency error"! We should either remove the circular dependancy or either have this solution. 

Just FYI because of how we do our imports (importing the whole feature and not the file) we have a lot of circular dependancies :) The problem occurs when we are defining stuff on the global scope of the file only because in this case they are `undefined`.

saghul commented on react/features/base/participants/middleware.js:
```diff
@@ -130,7 +136,13 @@ MiddlewareRegistry.register(store => next => action => {
-    return next(action);
+    const result = next(action);
+
+    if (type === SET_LOCATION_URL) {
```
Please don't add if checks if there is already a switch-case, just add another `case`!

hristoterezov replied:
This cannot go into the switch because I need the action to be executed. I think you've missed `const result = next(action);` maybe?

saghul commented on react/features/base/participants/middleware.js:
```diff
@@ -407,6 +419,26 @@ function _registerSounds({ dispatch }) {
+ * @private
+ * @returns {void}
+ */
+function _updateLocalParticipantFromUrl({ dispatch, getState }) {
```
I'm not sure about this change. base/settings takes care of user settings, and then it updates the local user with those. For example, when you configure the email in the settings menu. Now, you moved half of that elsewhere, so one needs to go to here to check that. It doesn't sound good.

The same way the config is overridden from the URL, some settings are (ie, this), that's why the code was where it was.

hristoterezov replied:
ok. I still think it doesn't belong here because it has nothing to do with the local settings but your explanation also makes sense. anyway I don't have strong preference about this and i'll revert the change...

hristoterezov commented:
Closing the PR due to @saghul 's comments! If we agree on something I'll just open another one for the issue with the circular dependancy.
